### DISCUSSION TITLE
linux-fs not found

### DISCUSSION THREAD

**Discussion by RonaldTHerman** [Pi4J/pi4j#330] [General]
My app requires 16 bit i2c register addresses.. I've extracted/tweaked the linus-fs code I need from AT24C512.java compiled and built.. when I run it,  the linux-fs provider can't be found. I've put the necessary files where I have the rest of the pi4j files. I'm missing something.

`Exception in thread "main" com.pi4j.provider.exception.ProviderNotFoundException: Pi4J provider [linuxfs-i2c] could not be found.  Please include this 'provider' JAR in the classpath.
        at com.pi4j.provider.impl.DefaultRuntimeProviders.get(DefaultRuntimeProviders.java:227)
        at com.pi4j.provider.impl.DefaultProviders.get(DefaultProviders.java:147)
        at com.pi4j.provider.Providers.get(Providers.java:253)
        at com.pi4j.context.Context.create(Context.java:327)
        at com.pi4j.internal.IOCreator.create(IOCreator.java:58)
        at com.pi4j.internal.IOCreator.create(IOCreator.java:126)
        at raspberrypi.I2C_fs.<init>(I2C_fs.java:77)
        at QWIK.RotaryEncoder.<init>(RotaryEncoder.java:56)
  `

The CLASSPATH/ library path (CLASSPATH=:/home/pi/NetBeansProjects/Tune/dist/lib) contains
`libpi4j-linuxfs.so     pi4j-core.jar                   pi4j-library-pigpio.jar        pi4j-plugin-raspberrypi.jar  Utilities.jar
log4j-api-2.20.0.jar   pi4j-library-linuxfs-2.1.1.jar  pi4j-plugin-linuxfs-2.1.1.jar  RaspberryPi.jar
log4j-core-2.20.0.jar  pi4j-library-pigpio-armhf.so    pi4j-plugin-pigpio.jar         slf4j-api-2.0.9.jar
` 

CONSTRUCTOR CODE:
`    public I2C_(int nAddr)
      {
             _pi4j = Pi4J.newAutoContext();
              _nAddr = nAddr;

               String id = String.format("Bus: 0X%02x: ", I2C_BUS);
               String name = String.format("Address: 0X%02x: ", nAddr);
		
		 _config = I2C.newConfigBuilder(_pi4j)
			.id(id)					// NEEDED??
			.bus(I2C_BUS)
			.device(_nAddr) 
			.name(name) 
			.provider("linuxfs-i2c")
			.build();
		 
		_i2c = _pi4j.create(_config);
		
		if (_i2c == null)
		{
			System.out.println(">>>>linuxfs-i2c Create failed");
			_isReady = false;
		}
		_isReady = true;
    }
	`

WHAT AM I MISSING

**Comment by RonaldTHerman** [comment]
Just to do an end-to-end check I changed "linuxfs-i2c" to pigpio-i2c and it found pigpio-i2c but got an IllegalStateException trying to do a writeRegister(). so I must have the path correct and I assume "linuxfs-i2c" is correct since I copied it from the EEPROM example.. What's different between  pigpio-i2c and linuxfs-i2c that prevents the latter from being found. I see "linuxfs-i2c" in LinuxFsI2CProvider.class so spelling must be correct.

 I'm using pi4j 2.5 SNAPSHOT.. I tried replacing pi4j-core.jar with pi4j-core-2.1.1 from the download but got a bunch of "cant find" symbol errors. I could really use some thoughts on this.

**Comment by taartspi** [comment]
I don't know your development environment.   When I build my devices with maven these jars are placed into target/distribution                                                                                                                                                                                           
jna-5.12.1.jar
jsch-0.1.55.jar
pi4j-core-2.5.0-SNAPSHOT.jar
pi4j-library-gpiod-2.5.0-SNAPSHOT.jar
pi4j-library-linuxfs-2.5.0-SNAPSHOT.jar
pi4j-library-pigpio-2.5.0-SNAPSHOT.jar
pi4j-plugin-gpiod-2.5.0-SNAPSHOT.jar
pi4j-plugin-linuxfs-2.5.0-SNAPSHOT.jar
pi4j-plugin-pigpio-2.5.0-SNAPSHOT.jar
pi4j-plugin-raspberrypi-2.5.0-SNAPSHOT.jar
pi4j-v2-devices-0.0.1.jar
pi4j-v2-devices-0.0.1-tests.jar
slf4j-api-1.7.30.jar
slf4j-simple-1.7.30.jar

**Reply by RonaldTHerman** [reply]
where do you find these downloads?  I'm not finding the SNAPSHOT files except for pi4j-2.5. And pi4j-v2-devices I don't see. Do I need plugin-gpio for just i2c? I've defined my own I2C-fs just for the device that requires 16 bit register addresses

**Reply by RonaldTHerman** [reply]
Another senior moment.. files were all hiding in the pi4j -2.5.0-SNAPSHOT download.. EXCEPT for pi4j-v2-devices-0.0.1.jar  I've never needed this jar.. what is it and where is a download?

**Reply by taartspi** [reply]
That jar contains my devices so present as i listed my environment. U do not need that jar

**Reply by RonaldTHerman** [reply]
good. In the meantime:

I built the app with the 2.5 SNAPSHOT files and ran it.. It is finding linuxfs-i2c but now I'm getting this exception

```
Exception in thread "main" com.pi4j.exception.Pi4JException: Failed to execute action for device 54 on bus 1
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CBus.executeIOCTL(LinuxFsI2CBus.java:106)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2C.readRegister(LinuxFsI2C.java:309)
        at com.pi4j.io.i2c.I2CRegisterDataReader.readRegister(I2CRegisterDataReader.java:139)
        at com.pi4j.io.i2c.I2CRegisterDataReader.readRegister(I2CRegisterDataReader.java:171)
        at raspberrypi.I2C_fs.readBytes(I2C_fs.java:143)
        at QWIK.RotaryEncoder.buttonPressed(RotaryEncoder.java:115)
        at tuner.Controller.ENCODER_main(Controller.java:598)
        at tuner.Controller.main(Controller.java:57)

Caused by: com.pi4j.library.linuxfs.LinuxFile$ScratchBufferOverrun: Scratch buffer overrun! Provide direct ByteBuffer for data larger than 2048 bytes
        at com.pi4j.library.linuxfs.LinuxFile.ioctl(LinuxFile.java:151)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CBus.executeIOCTL(LinuxFsI2CBus.java:90)
        ... 7 more
```

device 54 (0x36) is the correct address on  BUS_1. I'm only reading 8 bytes so where is the overrun?

The code in question is:

```
/**
* method to see if encoder button pressed
* @return true if pressed
*/
public boolean buttonPressed() {
  byte bfr[] = new byte[8];
  _i2c.readBytes(PINS, bfr, bfr.length);                      // PINS=0x00010004
  long ret = (bfr[4] << 24) | (bfr[5] << 16) | (bfr[6] << 8) | bfr[7];
        
  long bits = bytesToLong(bfr);
  _log.debug("bits=" + bits);
        
  boolean bPressed = (bits  & (1 << BTN_PIN)) != 0;
  _log.debug("bPressed= " + bPressed);
        
  return(bPressed);
}

/**
* method to read bytes from i2c
* @param register 16 bit address
* @param rData place to put data
* @param numBytes number of bytes to read
* @return  number of bytes read (?)
*/
public int readBytes(long register, byte[] rData, int numBytes) {
  byte[] regByte = new byte[2]; 
  regByte[0] = (byte) ((register & 0xff00) >> 8);
  regByte[1] = (byte) (register & 0xff);
  int rc = _i2c.readRegister(regByte, rData);
  return(rc);
}
```

**Reply by eitch** [reply]
Can you please properly format the code?

**Comment by RonaldTHerman** [comment]


**Reply by RonaldTHerman** [reply]
I didn't know there was a "proper" format.. I've been writing java code for years and this is the way I do it.. We'll have to agree to disagree. Also I'm not a frequent user of github so the format got  scrambled some when I posted it. Sorry about that.

**Reply by FDelporte** [reply]
@eitch was probably referring to the code snippet given above. please use three accents ``` to start and end a code block to make it easily readable here. I corrected it in your reply.

**Reply by RonaldTHerman** [reply]
Just found the help page for formatting. thanks for fixing it.

**Comment by RonaldTHerman** [comment]
Been fiddling around and I got rid of the exception for the time being. Following the Adafruit C code example I was reading 8 bytes from the device. I chanted it to 4 bytes and the exception went away. Still not getting cprrect values back though. Thanks for the help from all. Think I'll close this for now..

**Comment by taartspi** [comment]
Does your code read some smaller number of bytes and then try 8, hitting the bug.    By reading 1 byte then 8 I was able to duplicate what you encountered.  Initially and always 8 worked.             I did a pull-request of a fix for this.  If approved I will let you know when/how available, but sounds like you may not want it right away if you are making progress around the exception.

**Comment by RonaldTHerman** [comment]
Yes, all the reads are 4 bytes except this 8 bytes. The. Adafruit has an 8 byte collection of “pseudo” digital pins one of which flags the encoder push button.

**Comment by RonaldTHerman** [comment]
I'm progressing but now I get:

```
Exception in thread "main" com.pi4j.exception.Pi4JException: Failed to execute action for device 54 on bus 1
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CBus.execute(LinuxFsI2CBus.java:138)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2C.write(LinuxFsI2C.java:82)
        at com.pi4j.io.IODataWriter.write(IODataWriter.java:106)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2C.writeRegister(LinuxFsI2C.java:143)
        at com.pi4j.io.i2c.I2CRegisterDataWriter.writeRegister(I2CRegisterDataWriter.java:144)
        at raspberrypi.I2C_fs.writeBytes(I2C_fs.java:158)
        at QWIK.RotaryEncoder.setPosition(RotaryEncoder.java:109)
        at QWIK.RotaryEncoder.<init>(RotaryEncoder.java:63)
        at tuner.Controller.ENCODER_main(Controller.java:588)
        at tuner.Controller.main(Controller.java:57)
Caused by: java.io.IOException: Remote I/O error
        at java.base/java.io.RandomAccessFile.writeBytes(Native Method)
        at java.base/java.io.RandomAccessFile.write(RandomAccessFile.java:559)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2C.lambda$write$1(LinuxFsI2C.java:83)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CBus.execute(LinuxFsI2CBus.java:123)
```

using:
```
    public int  writeBytes(int register, byte[] wData, int numBytes)
	{
		byte[] regByte = new byte[2];
		splitRegAddr(register, regByte);  // splits 16 bit register into two bytes
                int rc = _i2c.writeRegister(regByte, wData, numBytes);
		_log.debug("writeRegister=" + rc);

        return(rc);
    }
```

this code was working at one point today.

**Reply by taartspi** [reply]
To make sure I am clear, this particular call to  writeRegister(regByte, wData, numBytes); I2CRegisterDataWriter.java:144, had functioned correctly some number of times and then began this failure ?

**Reply by RonaldTHerman** [reply]
That's correct but in the meantime I had made some small changes in other parts of my code.

**Reply by RonaldTHerman** [reply]
Well things were working today but getting wrong results. Then the remote I/O exception popped up again.. see above.
I'm pretty sure I'm talking to the Adafruit encoder correctly.

**Comment by taartspi** [comment]
1. Your problem changing from 4 to 8 byte read. There is an approved fix but it is NOT in a snapshot.  There is other i2c work i'd prefer to complete prior to a new snapshot. But if you need the fix speak up and I will ask for a new 2.5.x snapshot.                     2. The IO exception.  I also had a Pi begin to show the same for no apparent reason.  I had been debugging and killing programs and rebuilding.  I powered down and back up the Pi and it resumed proper operation, no exceptions.   At your convenience would you power cycle your Pi and see if the exception occurs afterwards.

**Reply by RonaldTHerman** [reply]
Powering down and up again sounds like an oldl IT solution :-).. I can do that later but can I just get the linuxfs jar files or is the whole pi4f distribution needed.? either way I'd like it. By the way this is runniing on a pi4j-4B.

**Reply by taartspi** [reply]
I did NOT do a snapshot of this fix so there is nothing to download. The other IT trouble report statement. "I didn't do anything, all i did was blah "

**Reply by taartspi** [reply]
I will ask for the snapshot. When available you should refresh all the snapshot 2.5 files you presently have in your workspace. I will tell you when available

**Reply by RonaldTHerman** [reply]
OK, I just powered up the Pi after dinner and get this

```
Exception in thread "main" com.pi4j.exception.Pi4JException: Failed to execute action for device 54 on bus 1
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CBus.executeIOCTL(LinuxFsI2CBus.java:106)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2C.readRegister(LinuxFsI2C.java:309)
        at com.pi4j.io.i2c.I2CRegisterDataReader.readRegister(I2CRegisterDataReader.java:139)
        at raspberrypi.I2C_fs.readBytes(I2C_fs.java:159)
        at QWIK.RotaryEncoder.getPin(RotaryEncoder.java:258)
        at QWIK.RotaryEncoder.buttonPressed(RotaryEncoder.java:191)
        at tuner.Controller.ENCODER_main(Controller.java:601)
        at tuner.Controller.main(Controller.java:57)
Caused by: com.pi4j.library.linuxfs.LinuxFile$ScratchBufferOverrun: Scratch buffer overrun! Provide direct ByteBuffer for data larger than 2048 bytes
        at com.pi4j.library.linuxfs.LinuxFile.ioctl(LinuxFile.java:151)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CBus.executeIOCTL(LinuxFsI2CBus.java:90)
```
fooling around after several runs of the app I got this

```
Exception in thread "main" com.pi4j.exception.Pi4JException: Failed to execute action for device 54 on bus 1
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CBus.executeIOCTL(LinuxFsI2CBus.java:106)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2C.readRegister(LinuxFsI2C.java:309)
        at com.pi4j.io.i2c.I2CRegisterDataReader.readRegister(I2CRegisterDataReader.java:139)
        at raspberrypi.I2C_fs.readBytes(I2C_fs.java:159)
        at QWIK.RotaryEncoder.getPosition(RotaryEncoder.java:122)
        at tuner.Controller.ENCODER_main(Controller.java:597)
        at tuner.Controller.main(Controller.java:57)
Caused by: com.pi4j.library.linuxfs.LinuxFile$LinuxFileException: Remote I/O error
        at com.pi4j.library.linuxfs.LinuxFile.ioctl(LinuxFile.java:172)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CBus.executeIOCTL(LinuxFsI2CBus.java:90)
```

**Comment by taartspi** [comment]
https://pi4j.com/download/              [pi4j-2.5.0-SNAPSHOT.zip (2.36 MiB, 2024-03-03 07:18)]              This build contains the fix for the error when you read from the I2C chip 4 bytes and then read 8 bytes.   Scratch buffer overrun

**Reply by RonaldTHerman** [reply]
Seems to be working but still getting garbage from Adafruit.. Also on that remote I/O exception...the one thing I hadn't tried is burning a new SD card.. I did that yesterday and now I haven't gotten that exception any more. Thanks for the 8 byte fix.. I need to quadruple check my Adafruit code.. One read I do (change in encoder position) appears to work but nothing else.. The 8 byte read is to get their pseudo gpio digital pins,  but in the doc they say 32 bit result but the code is 64 bit??

**Comment by RonaldTHerman** [comment]
Is there any documentation/source for linuxfs?  All I have is what Netbeans provides when it decompiles the library methods.
I'm finding that the methods using 2 byte register address "work" but the 16 bit integer versions don't giving
```
Exception in thread "main" com.pi4j.exception.Pi4JException: Failed to execute action for device 54 on bus 1
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CBus.execute(LinuxFsI2CBus.java:138)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2C.write(LinuxFsI2C.java:82)
        at com.pi4j.io.IODataWriter.write(IODataWriter.java:106)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2C.writeRegister(LinuxFsI2C.java:130)
        at com.pi4j.io.i2c.I2CRegisterDataWriter.writeRegister(I2CRegisterDataWriter.java:117)
        at raspberrypi.I2C_fs.writeBytes(I2C_fs.java:233)
        at raspberrypi.I2C_fs.writeBytes(I2C_fs.java:197)
        at QWIK.RotaryEncoder.writePin(RotaryEncoder.java:253)
        at QWIK.RotaryEncoder.turnLEDon(RotaryEncoder.java:224)
        at tuner.Controller.ENCODER_main(Controller.java:608)
        at tuner.Controller.main(Controller.java:57)
Caused by: java.io.IOException: Remote I/O error
        at java.base/java.io.RandomAccessFile.writeBytes(Native Method)
        at java.base/java.io.RandomAccessFile.write(RandomAccessFile.java:559)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2C.lambda$write$1(LinuxFsI2C.java:83)
        at com.pi4j.plugin.linuxfs.provider.i2c.LinuxFsI2CBus.execute(LinuxFsI2CBus.java:123)
        ... 10 more
```
Also Netbeans doesn't seem to recognize a offset parameter method call in the read/write bytes  methods. Without the offset it shows the parameter names for the method declaration parameters in light text but with offset it just shows "arg1, arg2" etc.  I could really use either source of method calls for the linuxfs code that is available for my use.

**Reply by FDelporte** [reply]
Sources are available here: https://github.com/Pi4J/pi4j-v2/tree/develop/plugins/pi4j-plugin-linuxfs/src/main/java/com/pi4j/plugin/linuxfs

**Comment by taartspi** [comment]
I  created PullRequest 333 an hour ago. Fixes two other linuxfs methods that are read from a specific register. U can hunt this up in the pi4j code repository for details  This change does not include the method in the call stack you sent yesterday.   Docs. There is no design doc specific to linuxfs. The couple of linuxfs java files in your call stack are the ones involved in runtime.  The pi4j web site snapshot downloads should have a jar of source code.  Next i was going to further investigate the exception in yesterday's call stack. It is the most simple code of the various methods, odd it has this intermittent exception

**Reply by RonaldTHerman** [reply]
I'm cornfused. My head is in a senior fog today.. It still is unclear whether my calls to read and write Register should include the "offset" parameter and if so what is it, always 0?

```
public int write(int register, byte[] data, int offset, int length)
public int read(int register,  byte[] data, int offset, int length)
```

I must be doing something correct. I have two i2c 4 byte reads to get current position and latest change. They both use the same code except for the 16 bit register. The change read returns seemingly correct values while the position read return one of two ridiculous values. I've checked and rechecked the register addresses and they are what Adafruit says to use. the common code is:
```
    public int readBytes(int register, byte[] rData, int numBytes)
     {
               byte[] regBytes = new byte[2]; // This is two byte register address
		splitRegAddr(register, regBytes);
		
		// make sure no garbage in array
		for (int i = 0 ; i < rData.length ; i++)
		{
			rData[i] = 0;
		}
                int rc = _i2c.readRegister(regBytes, rData, 0, numBytes);
		 if (rc < 0)
		 {
			 System.out.println("readRegister error");
			 _log.error("readRegister error");
		 }

                  return(rc);
         }
```
what do the read/write methods return.. I assumed it was number of bytes read/written but my 4 byte reads are returning "30".

Also in LinuxFsI2C.java it appears you are only using 1 byte for the register unless I am looking in the wrong place
```  
  public int writeRegister(int register, byte[] data, int offset, int length) {
        Objects.checkFromIndexSize(offset, length, data.length);
        byte[] tmp = new byte[length + 1];
        tmp[0] = (byte) register;  <<<<<<<<<<<<
        System.arraycopy(data, offset, tmp, 1, length);
        return write(tmp);
    }

    public int readRegister(int register, byte[] buffer, int offset, int length) {
        Objects.checkFromIndexSize(offset, length, buffer.length);
         byte reg[] = new byte[1];     <<<<<<<<<<<<
         reg[0] = (byte) (register & 0xff);
         int rCode = this.readRegister(reg, buffer, offset, length);
         return rCode;
    }
```

**Reply by RonaldTHerman** [reply]
1. what is the return for the read and write operations? byte count.. For 4 byte reads I get 30.
2. what should I set the offset parameter to, 0?

**Reply by taartspi** [reply]
So the removal of the initial write creating the ioexception answers my request. It prevents the ioexception. Leave your code with that initial write or a simple read().  I will figure out the real fix.    Offset. It is the beginning offset in the buffer for the data to write or the beginning offset to place the read data.  So often it is zero.    Specifically what method are you questioning the return value ?

**Reply by RonaldTHerman** [reply]
What do all read write methods return except for the single byte read which obviously  returns the desired byte.. Is the return some kind of error code or the number of bytes read or written. I get 30 from a four byte read.

**Reply by RonaldTHerman** [reply]
looking at your code
```
    @Override
    public int writeRegister(int register, byte b) {
        return write((byte) register, b); <<<<<<<<< b is the byte
    }

public int readRegister(int register) {
        byte reg[] = new byte[1];  <<<<<<<<< one byte
        reg[0] = (byte) (register & 0xff); <<<<<<<<<
        byte buffer[] = new byte[1];
        this.readRegister(reg, buffer,0,buffer.length);
        int rVal =  buffer[0] & 0xff;
        return rVal;
    }

```
This doesn't appear to be using a 16 bit register address unless I am missing something. I noticed this elsewhere as noted above.

**Reply by taartspi** [reply]
The i2c bus operates with 8 bit , and the devices data sheets use 8 bit(one byte) register address in their details.  If, as with your chip, the register requires more than one byte then a unique method is required.

**Reply by RonaldTHerman** [reply]
Yes and I thought that was what linuxfs was doing.. these examples taken from the source I was referred to for LinuxFsI2C.java do not use the entire 16 bit  register address..???

**Comment by taartspi** [comment]
[pi4j-2.5.0-SNAPSHOT.zip (2.36 MiB, 2024-03-10 15:23)](https://github.com/Pi4J/download/raw/main/pi4j-2.5.0-SNAPSHOT.zip)  That is the most recent linuxfs updates. These are Pigpio compatibility updates that may not effect you.  However I was able to create your ioexception.  Assuming  your first linuxfs call is the one using a two byte register address.  As an experiment can you place the simple linuxfs read() operation as your first linuxfs access.     Please let me know if that stops the ioexceptions ?

**Reply by RonaldTHerman** [reply]
I'm unable to recreate the original problem. In initializing the encoder I write1, read1, read4, read4, write4, 
 write4, write4, write4. The first 1 byte write is the reset for the encoder. If I omit that the first 4 byte write fails with the remote I/O exception

**Comment by taartspi** [comment]
Not certain which encoder you are using, I obtained an adafruit  I2C QT Rotary Encoder.  Maybe the same and it can help. 
    It arrived and   I coded up support using the adafruit python as docs, since this HW company won't release specs on the firmware they have running on the chip.
https://github.com/Pi4J/pi4j-example-devices/blob/master/src/main/java/com/pi4j/devices/rotary_encoder/RotaryEncoderId5880I2C.java

**Comment by RonaldTHerman** [comment]
I gave up on adafruit.. I used some eeprom code as a starting point. It required Linux-fs which is in the pi4j distribution. I gave up and now use sparkfun device. I’ve not seen the attached code. Maybe I’ll go back and try it. Thanks ---------------------------------Ronald T. ***@***.*** NHOn Aug 27, 2024, at 6:57 PM, Tom Aarts ***@***.***> wrote:﻿
Not certain which encoder you are using, I obtained an adafruit  I2C QT Rotary Encoder.  Maybe the same and it can help.
It arrived and   I coded up support using the adafruit python as docs, since this HW company won't release specs on the firmware they have running on the chip.
https://github.com/Pi4J/pi4j-example-devices/blob/master/src/main/java/com/pi4j/devices/rotary_encoder/RotaryEncoderId5880I2C.java

**Comment by IndianBartonka** [comment]
I found way to fix it you just need to build own context with this provider

```java
  final var pi4jBuilder = Pi4J.newContextBuilder();
        pi4jBuilder.add(new LinuxFsI2CProviderImpl());
        final Context pi4j = pi4jBuilder.build();
        ```

**Reply by RonaldTHerman** [reply]
Thanks but I, long ago found a way around this problem
